### ISSUE TITLE
No negative cosine similarities for IBCF with user mean-centered ratings

### ISSUE THREAD

**Issue by Daniel-Perruchoud** [mhahsler/recommenderlab#48]
Hello Michael
Thank you for a great package! I am using recommenderlab with my students since it nicely offers to experiment with many of the concepts published in the literature.
In the students' challenge we are struggling with the fact, that for the MovieLense data set no negative Item-Item similarities result when applying an IBCF recommender with "cosine" on user-mean centered ratings.
Most likely this is due to proxy::dist() but we would be more than happy if you could shed some light over what we might be doing wrong, if anything.
We have included a reproducible example below.

**Minimal example**

```
# Movie-to-Movie cosine similarity on user-centered MovieLense ratings
# Negative similarities seem to be mapped to positive similarities


library(recommenderlab)
library(tidyverse)

# minimal example
x <- rbind(c(1,-1),c(-1,1))
rating <- as(matrix(x, ncol = 2), "realRatingMatrix")

ibcf <- Recommender(rating, method = 'IBCF', 
                    parameter = list(method = 'cosine', k = dim(MovieLense)[2], 
                                     normalize = NULL, na_as_zero = FALSE))

as(ibcf@model$sim, "matrix")[1,2]
# --> 1 is not correct

proxy::dist(x = x, y = x, method = "cosine")[1,2]
# --> 0 is not correct

lsa::cosine(x[1,], x[2,])
# --> -1 is correct
```

**Movie Lense example**

_A) All movies considered_

--> no negative similarities

```
# MovieLense example 
data("MovieLense")

# @k: no neighborhood constraint 
ibcf <- Recommender(MovieLense, method = 'IBCF', 
                    parameter = list(method = 'cosine', k = dim(MovieLense)[2], 
                                     normalize = "center", na_as_zero = FALSE))

# similarity distribution
summary(as.vector(as(ibcf@model$sim, "matrix")))

#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.0000  0.0000  0.2668  0.4067  0.8992  1.0000 
# --> no negative similarities
```

_B) Two movies selected_

--> similarities match up to negative sign

```
# Example: "Money Train (1995)") vs "One Flew Over the Cuckoo's Nest (1975)"

# extract similarity
similarity1 <- as(ibcf@model$sim, "matrix")[536, 354]

# normalize ratings
center_MovieLense <- normalize(MovieLense, method = "center")
centered_ratings_twomovies <- as(center_MovieLense@data, "matrix")[, c(536, 354)]

# mimic nas_as_zero = FALSE
centered_ratings_twomovies[centered_ratings_twomovies[,1] == 0 | centered_ratings_twomovies[,2] == 0] <- NA 
centered_ratings_twomovies[is.na(centered_ratings_twomovies)] <- 0

similarity2 <- as.numeric(lsa::cosine(centered_ratings_twomovies[,1], centered_ratings_twomovies[,2]))


paste0("cosine similarity recommenderlab: ", similarity1)
# --> "cosine similarity recommenderlab: 0.660194486849794"

paste0("cosine similarity lsa: ", similarity2)
# --> "cosine similarity lsa: -0.660194486849794"

# --> similarities match up to negative sign

# check
# which(MovieLense@data@Dimnames[[2]] == "Money Train (1995)")
# which(MovieLense@data@Dimnames[[2]] == "One Flew Over the Cuckoo's Nest (1975)")
# which(ibcf@model$sim@Dimnames[[1]] == "Money Train (1995)")
# which(ibcf@model$sim@Dimnames[[1]] == "One Flew Over the Cuckoo's Nest (1975)")
```

_Session Info_
```
sessionInfo()

# R version 4.0.2 (2020-06-22)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 10 x64 (build 18363)
# 
# Matrix products: default
# 
# locale:
#   [1] LC_COLLATE=German_Switzerland.1252  LC_CTYPE=German_Switzerland.1252    LC_MONETARY=German_Switzerland.1252 LC_NUMERIC=C                       
# [5] LC_TIME=German_Switzerland.1252    
# 
# attached base packages:
#   [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#   [1] forcats_0.5.0        stringr_1.4.0        dplyr_1.0.2          purrr_0.3.4          readr_1.4.0          tidyr_1.1.2          tibble_3.1.0        
# [8] ggplot2_3.3.2        tidyverse_1.3.0      recommenderlab_0.2-7 registry_0.5-1       proxy_0.4-24         arules_1.6-6         Matrix_1.2-18       
# 
# loaded via a namespace (and not attached):
#   [1] Rcpp_1.0.5       lubridate_1.7.9  lattice_0.20-41  assertthat_0.2.1 digest_0.6.25    utf8_1.1.4       lsa_0.73.2       R6_2.4.1        
# [9] cellranger_1.1.0 backports_1.1.10 reprex_0.3.0     evaluate_0.14    httr_1.4.2       pillar_1.5.1     rlang_0.4.10     readxl_1.3.1    
# [17] rstudioapi_0.13  recosystem_0.4.3 irlba_2.3.3      blob_1.2.1       rmarkdown_2.4    labeling_0.3     munsell_0.5.0    broom_0.7.5     
# [25] compiler_4.0.2   modelr_0.1.8     xfun_0.18        pkgconfig_2.0.3  htmltools_0.5.0  tidyselect_1.1.0 fansi_0.4.1      crayon_1.3.4    
# [33] dbplyr_1.4.4     withr_2.3.0      SnowballC_0.7.0  grid_4.0.2       jsonlite_1.7.1   gtable_0.3.0     lifecycle_0.2.0  DBI_1.1.0       
# [41] magrittr_2.0.1   scales_1.1.1     cli_2.3.1        stringi_1.5.3    farver_2.0.3     fs_1.5.0         xml2_1.3.2       ellipsis_0.3.1  
# [49] generics_0.1.0   vctrs_0.3.4      tools_4.0.2      glue_1.4.2       hms_0.5.3        yaml_2.2.1       colorspace_1.4-1 rvest_0.3.6     
# [57] knitr_1.30       haven_2.3.1     
```

Best, Daniel

**Comment by mhahsler** [comment]
Hi Daniel,

thank you for the detailed examples. Let me talk with the maintainer of proxy to see what his take on the conversion from cosine values to distances/similarities is.

-Michael

**Comment by Daniel-Perruchoud** [comment]
Hi Michael

Thank you for the quick response! Looking forward to get your/their response.

Best regards from Switzerland,
Daniel

Von meinem iPad gesendet

Am 05.06.2021 um 18:22 schrieb Michael Hahsler ***@***.***>:

﻿

Hi Daniel,

thank you for the detailed examples. Let me talk with the maintainer of proxy to see what his take on the conversion from cosine values to distances/similarities is.

-Michael

**Comment by mhahsler** [comment]
The proxy maintainer has confirmed the bug and it will be fixed in the next release of proxy (Version 0.4-26). This should also fix the problem you are experiencing with recommenderlab.

**Comment by Daniel-Perruchoud** [comment]
Many thanks, Michael! Best regards, Daniel

Von meinem iPad gesendet

Am 07.06.2021 um 19:34 schrieb Michael Hahsler ***@***.***>:

﻿

The proxy maintainer has confirmed the bug and it will be fixed in the next release of proxy (Version 0.4-26). This should also fix the problem you are experiencing with recommenderlab.
